File: scripts/day8-part2-optimized.js
const fs = require("fs");

function distance(p1, p2) {
  const dx = p1[0] - p2[0];
  const dy = p1[1] - p2[1];
  const dz = p1[2] - p2[2];
  return Math.sqrt(dx * dx + dy * dy + dz * dz);
}

class UnionFind {
  constructor(n) {
    this.parent = Array.from({ length: n }, (_, i) => i);
    this.size = Array(n).fill(1);
    this.numComponents = n;
  }

  find(x) {
    if (this.parent[x] !== x) {
      this.parent[x] = this.find(this.parent[x]);
    }
    return this.parent[x];
  }

  union(x, y) {
    const rootX = this.find(x);
    const rootY = this.find(y);

    if (rootX === rootY) return false;

    if (this.size[rootX] < this.size[rootY]) {
      this.parent[rootX] = rootY;
      this.size[rootY] += this.size[rootX];
    } else {
      this.parent[rootY] = rootX;
      this.size[rootX] += this.size[rootY];
    }

    this.numComponents--;
    return true;
  }
}

function solvePart2(input) {
  const junctions = input.map((line) => line.split(",").map(Number));
  const n = junctions.length;

  // Create all pairs with their distances
  const pairs = [];
  for (let i = 0; i < n; i++) {
    for (let j = i + 1; j < n; j++) {
      pairs.push({
        i,
        j,
        dist: distance(junctions[i], junctions[j]),
      });
    }
  }

  // Sort by distance
  pairs.sort((a, b) => a.dist - b.dist);

  // Connect pairs until all are in one circuit
  const uf = new UnionFind(n);
  let lastConnection = null;

  for (const pair of pairs) {
    if (uf.union(pair.i, pair.j)) {
      lastConnection = pair;
      if (uf.numComponents === 1) {
        break;
      }
    }
  }

  // Return product of X coordinates
  return junctions[lastConnection.i][0] * junctions[lastConnection.j][0];
}

const input = fs
  .readFileSync("../inputs/day-8-input.txt", "utf-8")
  .trim()
  .split(/\r?\n/);

const iterations = 1;
const startTime = process.hrtime.bigint();

let result;
for (let i = 0; i < iterations; i++) {
  result = solvePart2(input);
}

const endTime = process.hrtime.bigint();
const totalNanos = Number(endTime - startTime);
const avgMicros = (totalNanos / iterations / 1000).toFixed(3);

console.log(`JavaScript (Day 8 Part 2 - Optimized)`);
console.log(`Answer: ${result}`);
console.log(`Average time: ${avgMicros} microseconds (${iterations} iterations)`);
console.log(`Total time: ${(totalNanos / 1000000).toFixed(2)} ms`);

File: scripts/day8-part2-optimized.py
import time
import math

def distance(p1, p2):
    dx = p1[0] - p2[0]
    dy = p1[1] - p2[1]
    dz = p1[2] - p2[2]
    return math.sqrt(dx * dx + dy * dy + dz * dz)

class UnionFind:
    def __init__(self, n):
        self.parent = list(range(n))
        self.size = [1] * n
        self.num_components = n
    
    def find(self, x):
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]
    
    def union(self, x, y):
        root_x = self.find(x)
        root_y = self.find(y)
        
        if root_x == root_y:
            return False
        
        if self.size[root_x] < self.size[root_y]:
            self.parent[root_x] = root_y
            self.size[root_y] += self.size[root_x]
        else:
            self.parent[root_y] = root_x
            self.size[root_x] += self.size[root_y]
        
        self.num_components -= 1
        return True

def solve_part2(input_data):
    junctions = [list(map(int, line.split(','))) for line in input_data]
    n = len(junctions)
    
    # Create all pairs with their distances
    pairs = []
    for i in range(n):
        for j in range(i + 1, n):
            pairs.append((distance(junctions[i], junctions[j]), i, j))
    
    # Sort by distance
    pairs.sort()
    
    # Connect pairs until all are in one circuit
    uf = UnionFind(n)
    last_connection = None
    
    for dist, i, j in pairs:
        if uf.union(i, j):
            last_connection = (i, j)
            if uf.num_components == 1:
                break
    
    # Return product of X coordinates
    return junctions[last_connection[0]][0] * junctions[last_connection[1]][0]

# Read input
with open("../inputs/day-8-input.txt", "r") as f:
    input_data = f.read().strip().split('\n')

# Benchmark
iterations = 1
start_time = time.perf_counter()

for _ in range(iterations):
    result = solve_part2(input_data)

end_time = time.perf_counter()
total_time = end_time - start_time
avg_time_us = (total_time / iterations) * 1_000_000

print(f"Python (Day 8 Part 2 - Optimized)")
print(f"Answer: {result}")
print(f"Average time: {avg_time_us:.3f} microseconds ({iterations} iterations)")
print(f"Total time: {total_time * 1000:.2f} ms")
